### PR TITLE
ddl: fix job state overridden when concurrent updates don't overlap in time range

### LINKED ISSUE #52747: DDL job state change might be overridden
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
below is the step when `cancel` operation is overridden. in real, any state change might be overridden on owner change cases where 2 owner might tries to change at same time, if the TXNs don't cross with each other, the prior one is overridden.
- apply below batch
- cancel the job before we start running it
```diff
diff --git a/pkg/ddl/ddl.go b/pkg/ddl/ddl.go
index 51fb9a9434..129f9cc00a 100644
--- a/pkg/ddl/ddl.go
+++ b/pkg/ddl/ddl.go
@@ -1724,6 +1724,8 @@ func resumePausedJob(_ *sess.Session, job *model.Job,
 	return nil
 }
 
+var GLOBAL = make(chan struct{})
+
 // processJobs command on the Job according to the process
 func processJobs(process func(*sess.Session, *model.Job, model.AdminCommandOperator) (err error),
 	sessCtx sessionctx.Context,
@@ -1794,6 +1796,8 @@ func processJobs(process func(*sess.Session, *model.Job, model.AdminCommandOpera
 		if err = ns.Commit(); err != nil {
 			continue
 		}
+		fmt.Println("after cancel job-------------------")
+		GLOBAL <- struct{}{}
 
 		for id, idx := range jobMap {
 			jobErrs[idx] = dbterror.ErrDDLJobNotFound.GenWithStackByArgs(id)
diff --git a/pkg/ddl/ddl_worker.go b/pkg/ddl/ddl_worker.go
index 6f6084fa61..dc84e45f05 100644
--- a/pkg/ddl/ddl_worker.go
+++ b/pkg/ddl/ddl_worker.go
@@ -853,6 +853,12 @@ func (w *worker) HandleJobDone(d *ddlCtx, job *model.Job, t *meta.Meta) error {
 }
 
 func (w *worker) prepareTxn(job *model.Job) (kv.Transaction, error) {
+	if job.Type == model.ActionTruncateTable && job.State == model.JobStateQueueing {
+		fmt.Println("before truncate -------------------", job.ID)
+		<-GLOBAL
+		time.Sleep(10 * time.Second)
+		fmt.Println("start truncate -------------------", job.ID)
+	}
 	err := w.sess.Begin()
 	if err != nil {
 		return nil, err

```

### 2. What did you expect to see? (Required)
job cancelled
### 3. What did you see instead (Required)
job cancel success, but the job keeps running as the meta is overridden
```sql
mysql> admin cancel ddl jobs 136;
+--------+------------+
| JOB_ID | RESULT     |
+--------+------------+
| 136    | successful |
+--------+------------+
1 row in set (0.00 sec)

mysql> admin show ddl jobs 1;
+--------+---------+------------+----------------+--------------+-----------+----------+-----------+---------------------+---------------------+---------------------+------------+
| JOB_ID | DB_NAME | TABLE_NAME | JOB_TYPE       | SCHEMA_STATE | SCHEMA_ID | TABLE_ID | ROW_COUNT | CREATE_TIME         | START_TIME          | END_TIME            | STATE      |
+--------+---------+------------+----------------+--------------+-----------+----------+-----------+---------------------+---------------------+---------------------+------------+
|    136 | test    | t          | truncate table | none         |         2 |      133 |         0 | 2024-04-18 23:30:30 | NULL                | NULL                | cancelling |
|    134 | test    | t          | truncate table | public       |         2 |      131 |         0 | 2024-04-18 23:26:27 | 2024-04-18 23:26:54 | 2024-04-18 23:30:40 | synced     |
+--------+---------+------------+----------------+--------------+-----------+----------+-----------+---------------------+---------------------+---------------------+------------+
2 rows in set (0.00 sec)

mysql> admin show ddl jobs 1;
+--------+---------+------------+----------------+--------------+-----------+----------+-----------+---------------------+---------------------+---------------------+--------+
| JOB_ID | DB_NAME | TABLE_NAME | JOB_TYPE       | SCHEMA_STATE | SCHEMA_ID | TABLE_ID | ROW_COUNT | CREATE_TIME         | START_TIME          | END_TIME            | STATE  |
+--------+---------+------------+----------------+--------------+-----------+----------+-----------+---------------------+---------------------+---------------------+--------+
|    136 | test    | t          | truncate table | public       |         2 |      133 |         0 | 2024-04-18 23:30:30 | 2024-04-18 23:31:34 | 2024-04-18 23:31:34 | synced |
+--------+---------+------------+----------------+--------------+-----------+----------+-----------+---------------------+---------------------+---------------------+--------+
1 row in set (0.01 sec)
```
### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: close #52747

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by wjhuang2016 — APPROVED** [review]

File: pkg/ddl/job_worker.go [hunk 1]
```diff
@@ -529,19 +530,32 @@ func (w *worker) prepareTxn(job *model.Job) (kv.Transaction, error) {
...
+	// time range of another concurrent job updates, such as 'cancel/pause' job
+	// or on owner change, overlap with us, we will report 'write conflict', but
+	// if they don't overlap, we query and check inside our txn to detect the conflict.
+	currBytes, err := sysTblMgr.GetJobBytesByIDWithSe(jobCtx.ctx, w.sess, job.ID)
+	if err != nil {
+		// TODO maybe we can unify where to rollback, they are scatting around.
+		w.sess.Rollback()
+		return 0, err
+	}
+	if !bytes.Equal(currBytes, jobW.Bytes) {
```
Comment (lance6716) [inline pkg/ddl/job_worker.go:554]: not sure what's the duration of comparing two 6M bytes.

**Review by lance6716 — APPROVED** [review]

File: pkg/ddl/job_worker.go (hunk 1 above)
Comment (D3Hunter) [reply pkg/ddl/job_worker.go:554]: worst case, about 210us
```
goos: darwin
goarch: arm64
pkg: try-out/test/misc
cpu: Apple M1
BenchmarkCompareBytes-8   	    6459	    211581 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	try-out/test/misc	1.408s
```
```go
func BenchmarkCompareBytes(b *testing.B) {
	size := 6 << 20
	l := make([]byte, 0, size)
	for i := 0; i < size; i++ {
		l = append(l, 'a')
	}
	r := make([]byte, 0, size)
	for i := 0; i < size; i++ {
		r = append(r, 'a')
	}

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_ = bytes.Equal(l, r)
	}
}
```

**Pull request merged**